jetty-quic/quic-common/src/main/java/org/eclipse/jetty/quic/common/QuicSession.java

```diff
@@ -0,0 +1,434 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.quic.common;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Connection;
+import org.eclipse.jetty.io.CyclicTimeout;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.quic.quiche.QuicheConnection;
+import org.eclipse.jetty.quic.quiche.QuicheConnectionId;
+import org.eclipse.jetty.quic.quiche.ffi.LibQuiche;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.component.LifeCycle;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.util.thread.ExecutionStrategy;
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Represents a logical connection with a remote peer, identified by a QUIC connection ID.</p>
+ * <p>Each QuicSession maintains a number of QUIC streams, identified by their QUIC stream ID;
+ * Each QUIC stream is wrapped in an {@link EndPoint}, namely {@link QuicStreamEndPoint}.</p>
+ * <p>Bytes received from a {@link QuicConnection} in {@link #process(SocketAddress, ByteBuffer)}
+ * are passed to Quiche for processing; in turn, Quiche produces a list of QUIC stream IDs that
+ * have pending I/O events, either read-ready or write-ready.</p>
+ * <p>On the receive side, a QuicSession <em>fans-out</em> to multiple {@link QuicStreamEndPoint}s.</p>
+ * <p>On the send side, many {@link QuicStreamEndPoint}s <em>fan-in</em> to a QuicSession.</p>
+ */
+public abstract class QuicSession
```

Comment:
@sbordet @lorban  I find the naming very confusion.   A `QuicSession` containers a `QuicheConnection` and the `QuicConnection` is not a connection at all, but more part of the `Connector` or maybe the `ConnectionFactory`.

I think it would be far simpler to rename `QuicSession` to `QuicConnection`, as that is what it is.  The current `QuicConnection` is only named that way because it is being shoehorned into the current SelectorManager and AbstractConnection.  It kind of fits, but the wrong name shows that there is probably a better way.